File: challenge-13/DistinctCount.java
/*
* Java Program to print the count of distinct elements
* in an input array
* Input Constraint: 5 <= arraySize <= 100
* 0 <= elementValue <= 100
*/

import java.util.Scanner;

public class DistinctCount {

	final int FREQUENCY_SIZE = 101;
	private int[] frequency = new int[FREQUENCY_SIZE];

	private int countDistinct(int[] array) {
		int index;
		int distinctCount = 0;

		// generate frequency table by incrementing each element's
		// frequency stored at index same as element
		for (int i = 0; i < array.length; i++) {

			index = array[i];
			frequency[index]++;
		}

		// If element occurs once, it's distinct
		for (int i = 0; i < FREQUENCY_SIZE; i++) {
			if (frequency[i] == 1) {
				distinctCount++;
			}
		}

		return distinctCount;
	}

	// Driver code
	public static void main(String[] args) {

		// Scanner object for input
		Scanner scanner = new Scanner(System.in);

		// Class object for accessing non-static members
		DistinctCount distinct = new DistinctCount();

		int arraySize;

		// Repeatedly input size until constraint satisfied
		do {
			System.out.print("Input size: ");
			arraySize = scanner.nextInt();
		} while (arraySize < 5 || arraySize > 100);

		int[] inputArray = new int[arraySize];

		System.out.println("Input array elements: ");

		// Input array, re-input out of bound values
		// by resetting i to previous index
		for (int i = 0; i < arraySize; i++) {
			System.out.print("Input element " + i +": ");
			inputArray[i] = scanner.nextInt();

			if (inputArray[i] > 100) {
				System.out.println("Error: Max Element Value = 100");
				i -= 1;
			} else if (inputArray[i] < 0) {
				System.out.println("Error: Min Element Value = 0");
				i -= 1;
			}
		}

		System.out.println("Distinct Count: " + distinct.countDistinct(inputArray));
	}
}

File: challenge-13/distinct_count.c
/*
* C Program to print the count of distinct elements
* in an input array
* Input Constraints: 5 <= size <= 100
* 0 <= element value <= 100
*/

#include <stdio.h>
#include <stdbool.h>

#define FREQUENCY_SIZE 101

// Returns distinct_count of distinct elements
int count_distinct(int array[], int size);

// Driver code
int main()
{
	int array_size;

	// Input till constraint satisfied
	do
	{
		printf("Input size: ");
		scanf("%d", &array_size);
	} while (array_size < 5 || array_size > 100);

	int input_array[array_size];

	printf("Input array elements: \n");

	// Input array, re-input out of bound elements
	for (int i = 0; i < array_size; i++)
	{
		printf("Input element %d: ", i);
		scanf("%d", &input_array[i]);

		// If element value out of bound, reset i to
		// previous value to input again
		if (input_array[i] < 0)
		{
			printf("\nError: Min Element Value = 0\n\n");
			i -= 1;
		}
		else if (input_array[i] > 100)
		{
			printf("\nError: Max Element Value = 100\n\n");
			i -= 1;
		}
	}

	printf("Distinct distinct_count: %d\n", count_distinct(input_array, array_size));

	return 0;
}

// Returns distinct_count of distinct elements
int count_distinct(int array[], int size)
{
	int index;
	int distinct_count = 0;
	// Declare and initialise all elements to 0
	int frequency[FREQUENCY_SIZE] = {};

	// Generate frequency of each element. Record frequency
	// of each element at its index
	for (int i = 0; i < size; i++)
	{
		index = array[i];
		frequency[index]++;
	} 

	// Count distinct, if frequency 1
	for (int i = 0; i < FREQUENCY_SIZE; i++)
	{
		if (frequency[i] == 1)
		{
			distinct_count++;
		}
	}

	return distinct_count;
}